### DISCUSSION TITLE
Link connection error

### DISCUSSION THREAD

**Discussion by reshin-pulapaddy** [clientIO/joint#2274] [Q&A]
### Introduction

When I connect links using the UI,  sometimes get an error such as "maximum call stack exceeded," which causes the program to freeze. and all link's beginning points are moved to the corner automatically. I get the same problem when I reload the page and click on the same link.
Error screenshot
![error 2](https://github.com/clientIO/joint/assets/117301232/3903058a-29b7-4d87-8bc1-ef390c19e51b)

Demo Video After reloading the page
![demo video](https://github.com/clientIO/joint/assets/117301232/c4b8af7e-4ae6-42fe-99ad-5f453fcd5a2e)

https://jsfiddle.net/reshin_pulapaddy/76ogyeLp/6/

### Does your question relate to JointJS or JointJS+. Select both if applicable.

JointJS, JointJS+

**Comment by kumilingus** [comment]
This is a minimal reproducible example:

```
L1 target === L2
L2 source === L1
```

<img width="353" alt="image" src="https://github.com/clientIO/joint/assets/3967880/bef6be96-c5fe-4e40-94bf-29501adfb027">


```js
graph.fromJSON({
    'cells': [
        {
            'type': 'app.Link',
            'router': {
                'name': 'normal'
            },
            'connector': {
                'name': 'rounded'
            },
            'labels': [],
            'source': {
                'x': 390,
                'y': 150
            },
            'target': {
                'id': 'ac4ff73b-2db3-446b-8311-f61230b0d9eb'
            },
            'id': 'faf6080d-0317-4f3f-b216-c26bad61c175',
            'z': 2,
            'vertices': [
                {
                    'x': 530,
                    'y': 250
                }
            ],
            'attrs': {}
        },
        {
            'type': 'app.Link',
            'router': {
                'name': 'normal'
            },
            'connector': {
                'name': 'rounded'
            },
            'labels': [],
            'source': {
                'id': 'faf6080d-0317-4f3f-b216-c26bad61c175'
            },
            'target': {
                'x': 570,
                'y': 430
            },
            'id': 'ac4ff73b-2db3-446b-8311-f61230b0d9eb',
            'z': 4,
            'vertices': [
                {
                    'x': 420,
                    'y': 330
                }
            ],
            'attrs': {}
        }
    ]
});
```
There are currently 2 issues:
1. It requires multiple updates to find the correct link shapes
2. `link.model.getBBox()` throws `Maximum call stack size exceeded`

---

What is the purpose of this type of link-to-link connection? What kind of diagram are you building? Could you share this information with us? 

There might be other ways how to achieve it and avoid this issue. Perhaps you could set the `validationConnection` function not to allow this type of loop connection.

I am trying to determine the priority of this bug. It doesn't seem to be a common use case. The actual fix might require a lot of time.

**Reply by reshin-pulapaddy** [reply]
Hi @kumilingus 

We are building an engineering drawing application.  In this application, we are considering links as pipes. To represent the joining of pipes, we need to connect two or more pipes together.
That's our business requirement Kindly let me know any alternative or ways to fix it as it's blocking our product usage.

Thank you

**Reply by reshin-pulapaddy** [reply]
Hi @kumilingus 
Hope you get a chance to look into this above requirement, could you please assist us with the solution?

**Reply by kumilingus** [reply]
Not yet. Please bear with me.

**Reply by reshin-pulapaddy** [reply]
Hi @kumilingus 

L1 target === L2
L2 source === L1

Are there any other scenarios in which the error might occur, or is this the only one?

**Reply by kumilingus** [reply]
There will be more scenarious.
<img width="528" alt="image" src="https://github.com/clientIO/joint/assets/3967880/049f545b-3e62-4b43-8b8a-286b8cc139dc">

**Reply by reshin-pulapaddy** [reply]
Hi @kumilingus 
![ezgif com-video-to-gif (1)](https://github.com/clientIO/joint/assets/117301232/d74ec9e4-5042-49f9-a2d7-1d76feb508fb)

When I drag a link towards another link, it should display the same highlighter, and when I release the mouse button, instead of joining the links, I need to correctly position the link end on top of the other link. How can we achieve this?

**Reply by kumilingus** [reply]
Yes, this is the kind of workaround I had in mind. When you attach a link to a link, you can create a new element at the connection point and attach the new link to it and split the original link into 2 parts (the first goes into the new element and the second goes out of it).

Or you can just set the link's target to a point (as you suggesting). This can be done on `link:connect` paper event. The connection point is available through the `linkView.targetPoint`.
```js
paper.on('link:connect', function(linkView) {
   linkView.model.target(linkView.targetPoint);
});
```

**Reply by reshin-pulapaddy** [reply]
@kumilingus 
The above code has been added to our project.
However, when I move one link to another, the highlighter appears. I need to correctly attach it on top of the link when I release the mouse in that highlighter displaying position.

![tteest](https://github.com/clientIO/joint/assets/117301232/835c2605-4193-43f6-aa16-a913e790eefd)

**Reply by kumilingus** [reply]
@reshin-pulapaddy please create a JSFiddle and I'll look at it.

**Reply by reshin-pulapaddy** [reply]
sure @kumilingus 

https://jsfiddle.net/reshin_pulapaddy/sp4L8zum/23/

**Reply by kumilingus** [reply]
Here's the updated code: https://jsfiddle.net/kumilingus/knfahvtq/

You need to find the closest point that lies on the link you were connecting to (I didn't realize that the link model is not connected to other link at `link:connect` 😕 ).
```js
paper.on('link:connect', function(linkView, evt, connectedLinkView) {
   linkView.model.target(connectedLinkView.getClosestPoint(linkView.targetPoint));
});
```

**Reply by reshin-pulapaddy** [reply]
Thank you @kumilingus, Now it's working

**Reply by reshin-pulapaddy** [reply]
Hi @kumilingus 
I am trying to implement the above functionality for other elements also

`this.paper.on('link:connect', (linkView, evt, connectedCellView, svg, end) => {
			if (connectedCellView?.model?.isLink()) {
				const point = connectedCellView.getClosestPoint(end === 'source' ? linkView.sourcePoint : linkView.targetPoint);
				linkView.model[end](point);
			} else {
				linkView.model[end](end === 'source' ? linkView.sourcePoint : linkView.targetPoint);
			}
		});`

`const paper = (this.paper = new joint.dia.Paper({
			width: this.backgroundImage.width,
			height: this.backgroundImage.height,
			gridSize: 5,
			drawGrid: true,
			model: graph,
			snapLinksSelf: true,
			restrictTranslate: true,
			cellViewNamespace: joint.shapes,
			autoResizePaper: false,
			defaultConnectionPoint: { name: 'boundary' },
			defaultLink: new joint.shapes.standard.Link({
				attrs: {
					line: {
						stroke: colorPalette.Grey,
						targetMarker: {
							type: 'path',
							d: 'M 0 0 0 0',
						},
					},
				},
			}),
			validateConnection: true,
			interactive: true,
			async: true,
			snapLabels: true,
			sorting: joint.dia.Paper.sorting.APPROX,
			connectionStrategy: joint.connectionStrategies.pinAbsolute,
			validateConnection: paperOptions.validateConnection,
		}));`

![snap](https://github.com/clientIO/joint/assets/117301232/18a4d355-3c3c-4f78-9579-622d38af8e10)


However, it connects the mouse-released point 
When I release the mouse on top of the element, it must attach at the midpoint of its side
how we can achieve it?

![image](https://github.com/clientIO/joint/assets/117301232/11177870-33e6-4032-92eb-352098e5938a)

**Reply by kumilingus** [reply]
You can check the `connectionStrategy` used in this [demo](https://codepen.io/jointjs/pen/KKGRJdG/c22433a3df81fedeaed0d8cc3ed71a14?editors=0010).
